infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java

```diff
@@ -654,20 +655,34 @@ public void accept(ProbeResult probeResult) {
   /** Listens container's events and terminates workspace if unrecoverable event occurs. */
   public class UnrecoverableEventHanler implements ContainerEventHandler {
     private List<String> events;
+    private Map<String, Pod> workspacePods;
+    private Date handlerInitialization;
 
-    public UnrecoverableEventHanler() {
+    public UnrecoverableEventHanler(Map<String, Pod> workspacePods) {
       this.events =
           Strings.isNullOrEmpty(unrecoverableEvents)
               ? Collections.EMPTY_LIST
               : Arrays.asList(unrecoverableEvents.split(","));
+      this.workspacePods = workspacePods;
+      this.handlerInitialization = new Date();
     }
 
+    /*
+     * Event is considered to be unrecoverable if it belongs to one of the workspace pods
+     * and 'lastTimestamp' of the event is *after* the the time of handler initialization
+     */
     @Override
     public void handle(ContainerEvent event) {
-      if (isUnrecoverable(event)) {
+      String podName = event.getPodName();
+      String eventLastTimestamp = event.getLastTimestamp();
+      Date eventLastTimestampDate = convertEventTimestampToDate(eventLastTimestamp);
+
+      if (isUnrecoverable(event)
+          && workspacePods.containsKey(podName)
+          && eventLastTimestampDate != null
+          && eventLastTimestampDate.after(handlerInitialization)) {
```

Comment:
I think it makes sense to perform this check in `KubernetesPods#watchContainers`. Then you'll even solve more issues like publishing to client old events by MachineLogsPublisher.

Reply:
+1

Reply:
@sleshchenko @garagatyi true, are you ok to add `Date getInitializationDate();` to `ContainerEventHandler` interface ?

Reply:
Sorry, but I don't understand why do you need it. I think you can save initialization date here https://github.com/eclipse/che/blob/ac3b4016492fcfb9361243bc708f1191d06c1adf/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesPods.java#L309 and do not publish events which happened before.

Reply:
so, you mean to check `Watcher` not `Handler` initialization date? or you mean to save Handler initialization date when adding handler to the `containerEventsHandlers` queue [1] ?

[1] https://github.com/eclipse/che/blob/ac3b4016492fcfb9361243bc708f1191d06c1adf/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesPods.java#L346

Reply:
> you mean to check Watcher not Handler initialization date?

yes. I think it would be good enough. We don't need any initialization data on Handler level. Handler only needs to handle **new** events. 

Offtop :smile: 
It is strange that we don't face the similar issue with `KubernetesPods#waitAsync` method. Maybe events are clear after pod removing or something like that.
Hmmm. I see. Everything is fine with `KubernetesPods#waitAsync` because of using a resourceVersion in BaseOperation.java 
```
    public Watch watch(Watcher<T> watcher) throws KubernetesClientException {
        return this.watch(this.resourceVersion, watcher);
    }
```
Looks like it doesn't work for all Events because we don't have a particular object with specified resourceVersion while subscribing.

Reply:
@sleshchenko well, I'm not sure that comparing `lastTimestamp` of event with Watcher initialization is a good idea. It would work currently since 2 handlers we have & watcher are initialized at roughly the same time. But imagine the situation that some new handler would be initialized differently - this would results in subtle bugs & need of rewriting watcher logic

Reply:
@ibuziuk 
Well, I expect that all old events will be processed and skipped immediately after establishing the WebSocket connection. So then:
1. The first handler is registered.
2. Watcher is initialized and subscribed, he receives all events that already happens and skips them. He publishes only events that happened after initialization.
3. The second handler is registered. It does not receive old events because a WebSocket connection is already established and all old events are already processed / skipped.

So nobody does not get an event that happened before handler initialization.

Do you mean an issue that there is a probability that the second handler will receive an event that occurred between watcher initialization and handler initialization? But why can it happen? Should a watcher receive an event just when it happened if a WebSocket connection is established?

Reply:
@sleshchenko ah, ok it makes total sense. let me rework to support this